### PR TITLE
Add converter for Kraken to BIOM.

### PR DESCRIPTION
**Pull request by davebx**
Concerns: For the tests to work, the JSON output is prettified. This results in somewhat larger output files.

### CODE REVIEW

**Comment by davebx** [comment]
Ping @jxtx due to expressed interest.

File: tools/kraken_to_biom/kraken_to_biom.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import os
+import datetime
+import subprocess
+
+
+def main( args ):
+    input_files = []
+    for filename, label in zip( args.files.split( ',' ), args.labels.split( ',' ) ):
+        fh, translated_filename = tempfile.mkstemp( suffix='.tab', prefix=filename )
+        os.close( fh )
```
Comment (bgruening) [inline tools/kraken_to_biom/kraken_to_biom.py]: fh.close()  I have not seen this before. Does this something special?

File: tools/kraken_to_biom/kraken_to_biom.xml [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        <![CDATA[
+        kraken_to_biom.py --files "$input_kraken" 
+            --labels "${','.join([dataset.name for dataset in $input_kraken])}"
+            --cluster $cluster
+            --output "$output_biom"
+            --taxonomy_path "${ncbi_taxonomy.fields.path}"
+        ]]>
+    </command>
+    <inputs>
+        <param argument="--files" format="tabular" label="Select two or more Kraken output files" min="2" multiple="True" name="input_kraken" type="data" />
```
Comment (bgruening) [inline tools/kraken_to_biom/kraken_to_biom.xml]: @blankenberg Your `filter by metadata header BLAST trick` would be handy here. Any progress on this front?

File: tools/kraken_to_biom/kraken_to_biom.xml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+<![CDATA[
+**What it does**
+
+This tool is designed to translate two or more results of the Kraken metagenomic classifier (see citations below) to the BIOM 1.0 format, enabling visualization with QIIME as well as other operations
+
+-------
+
+**Example**
+
+Given two Kraken output files like the following::
```
Comment (bgruening) [inline tools/kraken_to_biom/kraken_to_biom.xml]: With which parameter set?

**Comment by bgruening** [comment]
@davebx I feel not confident to review the converter but the wrappers looks great! So :+1 from me.

File: tools/kraken_to_biom/kraken_to_biom.py (hunk 1 above)
Comment (blankenberg) [inline tools/kraken_to_biom/kraken_to_biom.py]: using ```with``` and `tempfile.NamedTemporaryFile` would be better here

File: tools/kraken_to_biom/kraken_to_biom.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+import tempfile
+import json
+import os
+import datetime
+import subprocess
+
+
+def main( args ):
+    input_files = []
+    for filename, label in zip( args.files.split( ',' ), args.labels.split( ',' ) ):
```
Comment (blankenberg) [inline tools/kraken_to_biom/kraken_to_biom.py]: I like zip, but should we do some check on matching the lengths of the lists? 

Also, could you use an append parameter type here, iterate over the multiple input in the xml and then not need to worry about the splitting?

File: tools/kraken_to_biom/kraken_to_biom.xml (hunk 2 above)
Comment (blankenberg) [inline tools/kraken_to_biom/kraken_to_biom.xml]: @bgruening Could you elaborate?

File: tools/kraken_to_biom/kraken_to_biom.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+        fh, translated_filename = tempfile.mkstemp( suffix='.tab', prefix=filename )
+        os.close( fh )
+        with open( translated_filename, 'w' ) as fh:
```
Comment (blankenberg) [inline tools/kraken_to_biom/kraken_to_biom.py]: chance of race condition naming collsion, use `NamedTemporaryFile` directly inside a with instead

**Comment by blankenberg** [comment]
Suggest:
- add shebang to top of ```.py``` files.
- add ```__version__``` string constant and ```--version``` argument
- Move the converter code to a separate package repository, where installing it will `chmod +x` the `.py` using the xml tags. Should also be able to require eg a python package. Now, script can be called by name directly, without specifying interpreter
- Add options to allow overriding column selections, so that any 'tabular' file can be used as input (tabular to biom converter), defaults can be set to kraken columns. can also hide under advance options (your choice)

**Comment by blankenberg** [comment]
That being said, i am +1 on merging this, if additional time is not available for changes at this time.

File: tools/kraken_to_biom/kraken_to_biom.xml (hunk 2 above)
Comment (bgruening) [inline tools/kraken_to_biom/kraken_to_biom.xml]: As I visited PSU you worked on annotating tabular files with metadata, column headers and so on. We could filter by those here, isn't it?

**Comment by bgruening** [comment]
@davebx your call! Thanks Dan!

**Comment by bgruening** [comment]
ping @davebx

**Comment by blankenberg** [comment]
made a pr against @davebx https://github.com/davebx/tools-iuc/pull/4

**Comment by blankenberg** [comment]
One important issue is that we should be using the taxonomy found in the kraken database directories, not the separately maintained ncbi taxonomy.

**Comment by davebx** [comment]
I'm not sure this should be applied to the `master` branch.

**Comment by nsoranzo** [comment]
@davebx Can you fix the 6 flake8 errors in https://travis-ci.org/galaxyproject/tools-iuc/jobs/147195352#L443 ?

**Comment by shiltemann** [comment]
@davebx oof, this one is ancient, still relevant you think? We might be able to use this [kraken-biom bioconda package](https://bioconda.github.io/recipes/kraken-biom/README.html) to perform this task?

**Comment by bernt-matthias** [comment]
Closing this in favor of https://github.com/galaxyproject/tools-iuc/pull/4763

**Pull request closed without merging**